common/gl_util/obj.hpp

```diff
@@ -0,0 +1,259 @@
+#pragma once
+
+#define TINYOBJLOADER_IMPLEMENTATION
+#include "lib/tiny_obj_loader.h"
+
+#define STB_IMAGE_IMPLEMENTATION
+#include "lib/stb_image.h"
+
+namespace ILLIXR {
+
+	// Struct which is used for vertex attributes.
+	// Interleaves position/uv data inside one VBO.
+	struct vertex_t {
+		GLfloat position[3];
+		GLfloat uv[2];
+	};
+
+	// Struct for drawable debug objects (scenery, headset visualization, etc)
+	// Performs its own GL calls. Not indexed.
+	struct object_t {
+		GLuint vbo_handle;
+		GLuint num_triangles;
+		GLuint texture;
+		bool has_texture;
+
+		void Draw() {
+			glBindBuffer(GL_ARRAY_BUFFER, vbo_handle);
+			glEnableVertexAttribArray(0);
+			glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, sizeof(vertex_t), (void*)offsetof(vertex_t, position));
+			glEnableVertexAttribArray(1);
+			glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, sizeof(vertex_t), (void*)offsetof(vertex_t, uv));
+
+			if(has_texture){
+				glBindTexture(GL_TEXTURE_2D, texture);
+			}
+
+			glDrawArrays(GL_TRIANGLES, 0, num_triangles * 3);
+
+			if(has_texture){
+				glBindTexture(GL_TEXTURE_2D, 0);
+			}
+		}
+	};
+
+	// Represents a scene obtained from a single OBJ file.
+	// Multiple objects can reside within the OBJ file; each
+	// object can have a single diffuse texture. Multi-material
+	// objects not supported.
+    class ObjScene {
+		public:
+
+		// Default constructor for intialization.
+		// If the full constructor is not called,
+		// Draw() will never do anything.
+		ObjScene() {
+			successfully_loaded_model = false;
+		}
+
+		// obj_dir is the containing directory that should contain
+		// the OBJ file, along with any material files and textures.
+		//
+		// obj_filename is the actual .obj file to be loaded.
+		ObjScene(std::string obj_dir, std::string obj_filename) {
+
+			// If any of the following procedures fail to correctly load,
+			// we'll set this flag false (for the relevant operation)
+			successfully_loaded_model = true;
+			successfully_loaded_texture = true;
+
+			std::string warn, err;
+
+			if(obj_dir.back() != '/') {
+				obj_dir += '/';
+			}
+
+			std::string obj_file = obj_dir + obj_filename;
+
+			// We pass obj_dir as the last argument to LoadObj to let us load
+			// any material (.mtl) files associated with the .obj in the same directory.
+			bool success = tinyobj::LoadObj(&attrib, &shapes, &materials, &warn, &err, obj_file.c_str(), obj_dir.c_str());
+			if(!warn.empty()){
+#ifndef NDEBUG
+				std::cout << "[OBJ WARN] " << warn << std::endl;
+#endif
+			}
+			if(!err.empty()){
+				std::cout << "[OBJ ERROR] " << err << std::endl;
+				successfully_loaded_model = false;
+			}
+			if(!success){
+				std::cout << "[OBJ FATAL] Loading of " << obj_filename << " failed." << std::endl;
+				successfully_loaded_model = false;
```

Comment:
I didn't notice this before but these should be fatal. There isn't much point in continuing with the program if there is no app (gldemo) running. I think an `abort()` in both scenarios should do.

Rest looks good.

Reply:
Hm. Okay